# Code Submission

*Chen Chen
 21202636*

## Section 1

In [48]:
# import required packages here
import pandas as pd
import numpy as np
from itertools import permutations
import collections
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [49]:
# load the basket.csv file into a pandas DataFrame (df)
path = r"D:/UCD/classes/Data Management & Mining/Association Rules/data/"
filename = "basket.csv"
df = pd.read_csv(path + filename)

# Use df.shape to see the number of rows and columns in the df
print(df.shape)

(15206, 12)


In [50]:
# Use df.head() to see the first 8 rows of the df
print(df.head(8))

# Use df.info() to see the general information of the df
print(df.info())

   ID           0     1                   2       3    4    5    6    7    8  \
0   1  whole milk  eggs         salty snack     NaN  NaN  NaN  NaN  NaN  NaN   
1   2  whole milk  eggs         white bread  yogurt  NaN  NaN  NaN  NaN  NaN   
2   3  whole milk  eggs                 NaN     NaN  NaN  NaN  NaN  NaN  NaN   
3   4  whole milk  eggs                 NaN     NaN  NaN  NaN  NaN  NaN  NaN   
4   5  whole milk  eggs                 NaN     NaN  NaN  NaN  NaN  NaN  NaN   
5   6  whole milk  eggs          rolls/buns     NaN  NaN  NaN  NaN  NaN  NaN   
6   7  whole milk  eggs                 NaN     NaN  NaN  NaN  NaN  NaN  NaN   
7   8  whole milk  eggs  whipped/sour cream     NaN  NaN  NaN  NaN  NaN  NaN   

     9   10  
0  NaN  NaN  
1  NaN  NaN  
2  NaN  NaN  
3  NaN  NaN  
4  NaN  NaN  
5  NaN  NaN  
6  NaN  NaN  
7  NaN  NaN  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15206 entries, 0 to 15205
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype 
---  ---

In [51]:
# Set the index of the df to the Transaction_ID column
df.set_index("ID", inplace = True)

# Use df.head() to see the first 3 rows of the df
print(df.head(3))

# change the value types in all the dataframe columns to be string
df = df.astype("str")

             0     1            2       3    4    5    6    7    8    9   10
ID                                                                          
1   whole milk  eggs  salty snack     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2   whole milk  eggs  white bread  yogurt  NaN  NaN  NaN  NaN  NaN  NaN  NaN
3   whole milk  eggs          NaN     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN


In [52]:
# Use the pandas describe method to get an overview of the dataset
print(df.describe())

                 0           1      2      3      4      5      6      7  \
count        15206       15206  15206  15206  15206  15206  15206  15206   
unique         141         144    135    129    112     97     81     66   
top     whole milk  whole milk    nan    nan    nan    nan    nan    nan   
freq          1418        1128   9988  12808  14320  14729  14925  15008   

            8      9     10  
count   15206  15206  15206  
unique     36      2      2  
top       nan    nan    nan  
freq    15155  15205  15205  


In [53]:
# create a list of lists, where each list contains the items in a given transaction
transaction_list = df.values.tolist()

# view the first transaction in the list
print(transaction_list[0])

['whole milk', 'eggs', 'salty snack', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan']


In [54]:
# from your list of lists, create a flattened list that contains all items purchased
all_item_list = [item for transaction in transaction_list for item in transaction]

# view the legnth of the list
print(len(all_item_list))

167266


In [55]:
# create a list of unique items from the flattened list
unique_item_list = list(set(all_item_list))

# remove any elements that items of interest (like empty strings or nan values)
unique_item_list.remove('nan')

# print out how many unique items we have
print(len(unique_item_list))

148


In [56]:
# generate the itemset permutations up to 2-itemsets
rules = list(permutations(unique_item_list, 2))

# print out the number of rules
print(len(rules))

# print some of the elements of the list of rules
print(rules[:5])

21756
[('whisky', 'hair spray'), ('whisky', 'rubbing alcohol'), ('whisky', 'specialty fat'), ('whisky', 'artif. sweetener'), ('whisky', 'oil')]


In [57]:
# generate the frequency counts for each unique item in the dataset
freq_item = collections.Counter(all_item_list)

# show the most common
print(freq_item.most_common())

[('nan', 127345), ('whole milk', 3102), ('other vegetables', 2025), ('berries', 1848), ('rolls/buns', 1714), ('soda', 1507), ('yogurt', 1399), ('root vegetables', 1186), ('beef', 1158), ('tropical fruit', 1046), ('bottled water', 939), ('citrus fruit', 812), ('pastry', 788), ('eggs', 778), ('pip fruit', 740), ('chocolate', 736), ('shopping bags', 729), ('canned beer', 717), ('coffee', 715), ('bottled beer', 698), ('whipped/sour cream', 654), ('brown bread', 651), ('pork', 629), ('newspapers', 600), ('white bread', 549), ('cake', 544), ('jam', 544), ('butter', 534), ('fruit/vegetable juice', 521), ('margarine', 491), ('chicken', 456), ('grapes', 432), ('frozen vegetables', 426), ('cream cheese ', 358), ('napkins', 340), ('waffles', 305), ('onions', 304), ('salty snack', 285), ('long life bakery product', 269), ('sugar', 266), ('butter milk', 262), ('ham', 258), ('frozen meals', 249), ('beverages', 248), ('specialty chocolate', 246), ('misc. beverages', 234), ('oil', 228), ('ice cream', 

In [58]:
# one-hot encode the data and show the first 5 rows of the resulting df
# don't forget to drop the 'nan' value
encoder = TransactionEncoder().fit(transaction_list)
onehot = encoder.transform(transaction_list).astype("int")

df_onehot = pd.DataFrame(onehot, columns = encoder.columns_).drop("nan", axis = 1)
print(df_onehot)

       abrasive cleaner  artif. sweetener  baby cosmetics  bags  \
0                     0                 0               0     0   
1                     0                 0               0     0   
2                     0                 0               0     0   
3                     0                 0               0     0   
4                     0                 0               0     0   
...                 ...               ...             ...   ...   
15201                 0                 0               0     0   
15202                 0                 0               0     0   
15203                 0                 0               0     0   
15204                 0                 0               0     0   
15205                 0                 0               0     0   

       baking powder  bathroom cleaner  beef  berries  beverages  \
0                  0                 0     0        0          0   
1                  0                 0     0        0      

In [59]:
# Generate frequent itemsets with a minimum support of 0.5%
freq_itemsets = apriori(df_onehot, min_support = 0.005, use_colnames = True)

# view the itemsets sorted by most frequent
print(freq_itemsets.sort_values(by = ['support'], ascending = False))

      support                          itemsets
77   0.191109                      (whole milk)
45   0.128568                (other vegetables)
2    0.116336                         (berries)
59   0.108181                      (rolls/buns)
66   0.095226                            (soda)
..        ...                               ...
130  0.005195  (whole milk, whipped/sour cream)
50   0.005130                      (photo/film)
38   0.005130                      (mayonnaise)
79   0.005130                   (berries, beef)
105  0.005064                 (jam, whole milk)

[133 rows x 2 columns]


How many itemsets are generated for the following support thresholds?:
* 0.1%    -   762
* 0.5%    -   133
* 1%     -   68
* 10%     -   4

In [60]:
# generate association rules with a confidence threshold of 10%
rules_df = association_rules(freq_itemsets, metric = 'confidence', min_threshold = 0.1)

# how many rules are generated?
print(len(rules_df))

42


In [61]:
# print out the rules sorted by lift
print(rules_df.sort_values(by = ['lift'], ascending = False).drop(columns=['antecedent support', 'consequent support', 'conviction']))

             antecedents         consequents   support  confidence      lift  \
21                (eggs)        (whole milk)  0.015520    0.312169  1.633464   
25     (root vegetables)  (other vegetables)  0.013087    0.171997  1.337790   
26    (other vegetables)   (root vegetables)  0.013087    0.101790  1.337790   
28         (white bread)  (other vegetables)  0.006116    0.171587  1.334602   
9              (berries)            (yogurt)  0.012429    0.106840  1.203414   
10              (yogurt)           (berries)  0.012429    0.140000  1.203414   
2         (bottled beer)           (berries)  0.005590    0.123367  1.060442   
20                (eggs)  (other vegetables)  0.006708    0.134921  1.049413   
0                 (beef)  (other vegetables)  0.009930    0.133274  1.036610   
6               (pastry)           (berries)  0.005787    0.113256  0.973529   
11        (bottled beer)        (whole milk)  0.008089    0.178520  0.934126   
40         (white bread)        (whole m

## Section 2 - OPTION

This section details one of three options. For the other two options, see the assignment instructions document (*Instructions_Assignment_1.pdf*).

Your task is to implement the Apriori algorithm from "scratch" using the same dataset as Section 1.

You can read the data into a dataframe and use pandas methods but you cannot use mlxtend or another similar package. You should write functions to implement the steps of the algorithm to create frequent itemsets.

Calculate the support, confidence, and lift for single itemset rules.

In [62]:
# import required packages
import pandas as pd
import numpy as np
from itertools import permutations
import collections

In [63]:
# load the basket.csv file into a pandas DataFrame (df)
path = r"D:/UCD/classes/Data Management & Mining/Association Rules/data/"
filename = "basket.csv"
df = pd.read_csv(path + filename)

In [64]:
# Set the thresholds for support and confidence
min_support = 0.005
min_confidence = 0.1

In [65]:
# Set the index of the df to the Transaction_ID column
df.set_index("ID", inplace = True)

In [66]:
# create a list of lists, where each list contains the items in a given transaction
transaction_list = df.values.tolist()

In [67]:
# from your list of lists, create a flattened list that contains all items purchased
all_item_list = [item for transaction in transaction_list for item in transaction]

In [68]:
# create a list of unique items from the flattened list
unique_item_list = list(set(all_item_list))

# remove any elements that items of interest (like empty strings or nan values)
unique_item_list.remove(np.nan)

## Step 1: Get the frequent itemsets for one item

In [69]:
# generate the frequency counts for each unique item in the dataset
item_counts = collections.Counter(all_item_list).most_common()

# Transform the frequency counts into a pandas dataframe
item_counts = pd.DataFrame(item_counts, columns = ["item", "counts"]).dropna()

# Calculate the support values
total_number = len(transaction_list)
item_counts["support"] = item_counts["counts"]/total_number

# Filter the frequent itemsets to meet the minimum support requirement
freq_itemsets = item_counts.loc[item_counts["support"]>=min_support]
freq_itemsets = freq_itemsets["item"]

## Step 2: Generate rules based on the frequent itemsets

In [70]:
# generate the itemset permutations up to 2-itemsets
rules = list(permutations(freq_itemsets, 2))

In [ ]:
# count the number of occurences for the rules
rule_counts = {}
flag = 0

for i in rules:
    for j in transaction_list:
        if set(i).issubset(set(j)):
            flag += 1
    rule_counts[i] = flag
    flag = 0

# Transform the rule counts into a pandas dataframe
rule_counts = pd.DataFrame.from_dict(rule_counts, orient = "index").reset_index()

In [ ]:
# get the antecedents and the consequents
rule_counts["antecedent"] = rule_counts["index"].apply(lambda x: x[0])
rule_counts["consequent"] = rule_counts["index"].apply(lambda x: x[1])

In [ ]:
# merge dataframes to get the counts for the antecedents and the consequents
temp = rule_counts.merge(item_counts, how = "left", left_on = "antecedent", right_on = "item")
result = temp.merge(item_counts, how = "left", left_on = "consequent", right_on = "item")

In [ ]:
# change the column names and drop unnecessary columns
result.rename(columns = {"index":"rule", 0:"rule_counts", "counts_x":"antecedent_counts", "counts_y":"consequent_counts", "support_x":"support"}, inplace = True)
result.drop(columns = ["item_x", "item_y", "support_y"], inplace = True)

In [ ]:
# Calculate the confidence values and the lift values
result["confidence"] = result["rule_counts"]/result["antecedent_counts"]
result["lift"] = result["confidence"]/(result["consequent_counts"]/total_number)

# Filter the result to show those with confidence>=10%
result = result[result["confidence"]>=0.1]

# sort the result by the lift values
print(result.sort_values(by = ["lift"], ascending = False)[:20])